Instrument python code for tracing

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/862
<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
Instrument python code for tracing

We're manually instrumenting our code for the moment, since most of our
python code isn't using frameworks. This does rely on setting an
endpoint via env vars, which should make operations easier between
local and AWS
<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
From the larger local-tracing branch.
The automated test suite should ensure that this doesn't break the code even without a true endpoint
<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Instrument python code for tracing

We're manually instrumenting our code for the moment, since most of our
python code isn't using frameworks. This does rely on setting an
endpoint via env vars, which should make operations easier between
local and AWS

Commit:
Remove unused import

Commit:
Don't upgrade pants to rc version

It appears that this pants upgrade may be breaking e2e tests for some
reason. See https://github.com/grapl-security/grapl/pull/1514

Commit:
Fix dependencies correctly

christophermaier commented on 3rdparty/python/requirements.txt:
```diff
@@ -1,13 +1,20 @@
 # Any Python dependencies used by our code should be declared here.
 # !!!!! Don't forget to regenerate the constraints from this! (see readme.md)
 
+# Please keep this alphabetical order
```
🙏 🙏 🙏

christophermaier commented on src/python/analyzer_executor/src/run.py:
```diff
@@ -23,28 +25,31 @@ async def _main() -> None:
-            await timeout_manager.keep_alive()
-        except SqsTimeoutManagerException:
-            LOGGER.error("SQS Timeout Manager exception", exc_info=True)
+    with TRACER.start_as_current_span("_main"):
```
Might be better to use `_main.__name__` instead of `"_main"` for this span name, just to guard against refactoring-induced errors in the future.

christophermaier commented on src/python/engagement-creator/engagement_creator/run.py:
```diff
@@ -5,41 +5,44 @@
-            await timeout_manager.keep_alive()
-        except SqsTimeoutManagerException:
-            LOGGER.error("SQS Timeout Manager exception", exc_info=True)
+    with TRACER.start_as_current_span("engagement_creator"):
```
Why is this `"engagement_creator"`, rather than `"_main"`, as it is above?

christophermaier commented on src/python/grapl-common/grapl_common/grapl_tracer.py:
```diff
@@ -0,0 +1,54 @@
+        # local_node_ipv6="2001:db8::c001",
+        # local_node_port=31313,
+        # max_tag_value_length=256
+        # timeout=5 (in seconds)
```
Are these additional arguments to eventually add later?

twunderlich-grapl replied:
These were the default arguments. I'm deleting them since I don't think we'll be adjusting them in here. More likely, we'll control opentelemetry with environment variables going forward

christophermaier commented on src/python/provisioner/provisioner/app.py:
```diff
@@ -98,22 +101,23 @@ def _retrieve_test_user_password(
-        role="owner",
-    )
-    LOGGER.info("created test user")
+    with TRACER.start_as_current_span("provision"):
```
Same comments here re: span name.

christophermaier commented on src/python/grapl-common/grapl_common/grapl_tracer.py:
```diff
@@ -0,0 +1,54 @@
+    provider.add_span_processor(processor)
+    trace.set_tracer_provider(provider)
+
+    BotocoreInstrumentor().instrument()
```
As this code stands, this is basically just going to instrument all AWS calls, right?

(I have no issue with that, to be clear; just clarifying my understanding!)

twunderlich-grapl replied:
Correct. Each AWS call will be a new span in the trace

christophermaier commented on src/python/grapl-common/grapl_common/grapl_tracer.py:
```diff
@@ -0,0 +1,54 @@
+        resource=Resource.create(
+            {
+                "service.name": service_name,
+                # TODO add things like instance_id, etc here
```
Will we eventually have some kind of schema for what kind of information we want to capture here? Or is there already some commonly-agreed-upon schema out in the wild that we could use (or are currently using in this code)?

twunderlich-grapl replied:
Yep. We'll be primarily using the opentracing schemas, although I could see us adding a few things that are grapl unique
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md

christophermaier reviewed (approved):
I have some relatively minor questions / suggestions, but otherwise looks good!

Commit:
Fix dependencies?

Commit:
Merge remote-tracking branch 'origin/main' into instrument-python

Commit:
Remove commented default arguments. These will be managed via environmental variables going forward

Commit:
Merge remote-tracking branch 'origin/main' into instrument-python

Commit:
Change span name to method/function name

Span should probably be the method name since we already have the
service and module names.

Commit:
Clean up comments

twunderlich-grapl commented:
@christophermaier 
I've set the span name to always be the method/function name via `__name__`. I think that should be the pattern going forward.

Commit:
Merge remote-tracking branch 'origin/main' into instrument-python

Commit:
Set tracing endpoints via env vars

Commit:
Merge remote-tracking branch 'origin/main' into instrument-python

Commit:
Bump requirements

For whatever reason, bumping dependencies seems to fix e2e tests. I
don't really think that could be the real reason why e2e tests were
failing but I'm willing to try it.
